### PR TITLE
fix(desktop,host-service): tolerate slow or failing post-checkout hooks during worktree creation

### PR DESCRIPTION
**Pull request by Kitenite**
## Summary
- Workspace creation no longer fails when a repo's post-checkout hook exits non-zero or runs long: if the worktree is verifiably in place (registered + valid HEAD/branch), the hook error is demoted to a warning and creation succeeds.
- Worktree-add / PR-checkout timeouts raised from 2 min to 10 min so hooks that install dependencies on fresh worktrees can finish.
- The tolerance helper moved to `@superset/shared/git-hook-tolerance` and now also guards the host-service (v2) `addBranchWorktree` path, so both creation surfaces behave identically.

## Why / Context
A user (drew) hit "Workspace setup failed" on every creation attempt in a repo whose post-checkout hook installs dependencies and builds packages (`Fresh worktree detected — installing dependencies… Done in 1m 5.7s Building packages…`). Two compounding bugs:

1. **The 120s exec timeout killed git mid-hook.** Checkout of ~11k files + a 1m+ install blows past 120s; Node SIGTERMs git and creation reports failure — with the worktree fully checked out on disk.
2. **The existing hook tolerance (SUPER-594) was gated backwards.** It pattern-matched the error text for the literal string `post-checkout` before checking whether the worktree actually exists. A timeout kill (or any hook that doesn't print that marker) never engages it.

Each failure stranded a fully valid worktree + branch, so retries failed on "already exists". These failures are also invisible in telemetry: `updateProgress("failed")` is never captured, only thrown tRPC errors are — the sibling PR-checkout path shows the same `fatal: '<path>' already exists` loop in PostHog (one user hit it 42×).

## How It Works
- `runWithPostCheckoutHookTolerance` (now in `@superset/shared/git-hook-tolerance`) consults `didSucceed()` on **any** error and tolerates when ground truth confirms the operation landed. The `post-checkout` text gate is gone.
- Desktop `execWorktreeAdd`: ground truth = worktree registered **and** `rev-parse --verify HEAD` resolves (a half-created worktree can stay registered); default timeout 600s. Same bump for `gh pr checkout` and its `FETCH_HEAD` fallback, which also run hooks.
- Host-service `addBranchWorktree`: both worktree-add variants wrapped; ground truth = existing `findWorktreeAtPath(git, worktreePath, plan.branch)` (registered at path with the planned branch — stricter than the desktop probe; this path never detaches). No timeout change needed there: simple-git never had one, so slow hooks were never killed on v2.

## Manual QA (CDP, dev desktop from this branch, before/after evidence gate)
Repro repo with a switchable post-checkout hook; "before" = this fix stashed, "after" = this branch. Desktop runs on the v1 surface (drew's surface — the error format `Command failed: git -C …` is the desktop execFile path).

**Before (reproduces both report screenshots):**
- [x] Hook exits non-zero (no "post-checkout" marker in output) → `step: failed` in 2s; error text matches the user report byte-for-byte; failure screen pixel-matches the screenshot
- [x] Hook sleeps 130s → `step: failed` at 122s (timeout kill), output truncated mid-hook — the second report screenshot
- [x] Ground truth both times: `git worktree list` shows a fully valid worktree with the branch checked out — the app failed anyway and stranded it

**After (same scenarios, same observations):**
- [x] Hook exits non-zero → `step: ready` in 2s
- [x] Hook sleeps 130s → `step: ready` at 132s (hook simply completes inside the new ceiling; no tolerance needed)
- [x] Stranded before-workspace recovers via a real-UI "Retry Setup" click → `ready` in 2s, workspace fully functional (old code looped on "already exists" here)
- [x] Main-process log shows the tolerance warning firing exactly where expected, and nowhere else
- [x] Host-service v2 E2E (post-refactor): `project.create` + `workspaces.create` via host tRPC against the failing-hook repo → workspace created, worktree registered with the planned branch; `[hs:…] … (non-fatal)` log line proves the shared helper ran in the host-service process

Creation was triggered via the same tRPC mutations the UI forms call (the folder picker is a native dialog CDP can't drive); navigation, the failure screen, the Retry Setup click, and screenshots were real UI.

## Testing
- `bun run typecheck` (repo-wide, exit 0)
- `bun run lint` (exit 0)
- `bun test` on the three touched suites: 44 pass — desktop `git.test.ts` (4 new edge cases: no-marker hook failure, git SIGTERM'd mid-hook with worktree intact, invalid start point still throws, pre-existing plain dir still throws), shared `git-hook-tolerance.test.ts`, new host-service `worktree-add-tolerance.test.ts` (new-branch + existing-branch tolerance, invalid start point)
- Mutation-verified: re-adding the old text-pattern gate fails the timeout-shape test; disabling `didSucceed` fails exactly the two host-service tolerance tests

## Design Decisions
- **Ground truth over pattern matching**: error-text heuristics can never enumerate every hook failure shape (timeout kills carry no marker at all). Checking the worktree's actual state is exact, and conservative in the right direction — when the probe itself fails, the original error is rethrown.
- **Share the policy, not the git clients**: desktop (shell-PATH execFile) and host-service (simple-git factory with credential/env handling) keep their own plumbing; with v1 sunsetting, unifying the exec layers is churn with no user benefit. Only the decision logic is shared.
- **Exported `addBranchWorktree`/`BranchSourcePlan`** so the host-service behavior is testable against real git repos.

## Known Limitations
- A timeout kill can orphan the hook's child processes (e.g. a package install continuing in the background). The 10-min ceiling makes this rare; creation now succeeds regardless.
- Stranded **detached** worktrees (PR flow killed mid-hook before `gh pr checkout` ran) do not self-heal on retry: a detached leftover is indistinguishable from a foreign checkout, so the "already exists" error surfaces — same as before this PR. Branch-add strandings (the common case) do recover. Safe PR-retry self-healing needs a provenance marker; follow-up.
- If the timeout kills git *before* checkout completes, nothing is registered, `didSucceed` is false, and the error surfaces as before — genuine failures still fail.

## Follow-ups
- Move user hooks out of the creation critical path entirely (`core.hooksPath=` during add, then run the hook as a streamed, non-fatal setup step).
- `workspace_init_failed` telemetry — these failures are currently invisible; this bug was only found via user screenshots.
- Clean up stranded worktrees/branches from historical failed attempts and make plain Retry self-heal (dedup at creation time instead of the post-failure button).
- Humanize the failure view: map common git fatals to one-line actionable messages; the raw command spew is line-clamped to 3 lines, which hides the actual `fatal:` line.

---
## Summary by cubic
Tolerates slow or failing post-checkout hooks during worktree creation and PR checkout by verifying ground truth (registered worktree, expected branch, resolvable `HEAD`, and PR head OID) so setup succeeds when checkout actually completed. Timeouts are raised to 600s, and the logic is shared via `@superset/shared/git-hook-tolerance` across desktop and host-service.

- **Bug Fixes**
  - Demote errors only when ground truth confirms success:
    - Branch adds: worktree is registered at the path with the requested branch; `HEAD` resolves. Detached adds only count when this call registered the path (detached retries are not tolerated).
    - PR checkout: tolerate only when the local branch is checked out and `HEAD` equals the PR’s `headRefOid` (covers `gh pr checkout` and the `FETCH_HEAD` fallback).

- **Refactors**
  - Moved tolerance helper to `@superset/shared/git-hook-tolerance` and adopted in desktop and host-service; host-service now also requires a resolvable `HEAD`.
  - Exported `addBranchWorktree` and `BranchSourcePlan`; added targeted tests in desktop, host-service, and shared packages.

<sup>Written for commit ca4054d0ea45bd6ca8e5277c75168b3f4ed1fcc7. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/superset-sh/superset/pull/6168?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Worktree creation now succeeds when post-checkout hooks fail after the worktree is otherwise valid.
  * Supports new branches, existing branches, detached worktrees, and pull request checkouts.
  * Longer operations have increased time limits.

* **Bug Fixes**
  * Prevents incomplete or mismatched worktrees from being marked successful.
  * Invalid references and conflicting worktrees continue to fail safely.
  * Successful operations with hook failures now show non-fatal warnings.

### CODE REVIEW

File: apps/desktop/src/lib/trpc/routers/workspaces/utils/git.ts [hunk 1]
```diff
@@ -161,8 +164,22 @@ async function execWorktreeAdd({
...
+				// resolvable HEAD to prove the checkout completed.
+				await execGitWithShellPath(
+					["-C", worktreePath, "rev-parse", "--verify", "HEAD"],
+					{ timeout: 10_000 },
+				);
+				return true;
+			} catch {
+				return false;
+			}
+		},
```
[the remark this answers is not in the document]
Comment (Kitenite) [reply apps/desktop/src/lib/trpc/routers/workspaces/utils/git.ts:206]: Fixed in 39b4ff05b. `execWorktreeAdd` now takes `expectedBranch`: branch adds only tolerate a failure when the requested branch is verifiably checked out at the path (pre-existing worktree on a different branch rethrows — covered by a new test, mutation-checked), and the detached PR variant only tolerates when this call itself registered the worktree (pre-registered paths rethrow, so `gh pr checkout --force` never runs in a foreign worktree). A same-branch pre-existing worktree still tolerates intentionally — that's the retry-after-stranded-attempt recovery path, also covered by a new test.

**Comment by Kitenite** [comment]
Re CodeRabbit's outside-diff comment (hook tolerance for PR checkout): fixed in 39b4ff05b — `gh pr checkout` and the FETCH_HEAD fallback now run under `checkoutBranchWithHookTolerance`, suppressing an error only when `localBranchName` is verifiably the current branch in the worktree. Not unit-tested directly (requires an authenticated `gh` + a live PR, which the test environment doesn't have); the tolerance predicate itself is the same `getCurrentBranch === target` path already covered by the shared-helper and branch-flow tests.

File: packages/shared/src/git-hook-tolerance.ts [hunk 2]
```diff
@@ -19,19 +19,14 @@ function getErrorText(error: unknown): string {
...
-		text.includes("command not found")
-	);
-}
-
+/**
+ * Runs a git checkout-ish command whose exit status can lie about the
+ * outcome: post-checkout hooks run after the checkout itself, so a hook
+ * that exits non-zero (or outlives the command timeout and gets killed)
+ * fails the command even though the worktree/branch is fully in place.
+ * `didSucceed` is the ground truth — when it confirms the operation
```
[the remark this answers is not in the document]
Comment (Kitenite) [reply packages/shared/src/git-hook-tolerance.ts:27]: Addressed across 39b4ff05b and cf2bd257b: every caller's predicate now verifies the requested ref — desktop branch adds require the requested branch checked out at the path (`expectedBranch`), the detached PR add only tolerates when this call registered the worktree, branch switches already compared against the target branch, and host-service requires the planned branch plus a resolvable HEAD. Verified E2E over CDP: a pre-existing worktree on a different branch fails creation and is left untouched.

File: packages/host-service/src/trpc/router/workspaces/workspaces.ts [hunk 3]
```diff
@@ -324,18 +326,31 @@ function isBranchInUseByWorktreeError(err: unknown): boolean {
...
+	// non-zero fails `worktree add` with the worktree fully in place. Every
+	// branch case below checks out `plan.branch`, so registered-at-path with
+	// that branch is the ground truth.
+	const runWorktreeAdd = (addArgs: string[]) =>
+		runWithPostCheckoutHookTolerance({
+			context: `Worktree created at ${worktreePath}`,
+			run: async () => {
+				await git.raw(addArgs);
+			},
+			didSucceed: () => findWorktreeAtPath(git, worktreePath, plan.branch),
```
[the remark this answers is not in the document]
Comment (Kitenite) [reply packages/host-service/src/trpc/router/workspaces/workspaces.ts]: Valid — fixed in cf2bd257b. `didSucceed` now also runs `rev-parse --verify HEAD` inside the worktree, matching the desktop contract (`getWorktreeBranchAtPath` parses `worktree list` output and never probes the worktree itself, so registration + listed branch alone could accept a half-created worktree).

File: apps/desktop/src/lib/trpc/routers/workspaces/utils/git.test.ts [hunk 4]
```diff
@@ -440,6 +440,120 @@ describe("createWorktree hook tolerance", () => {
...
+		expect(existsSync(worktreePath)).toBe(true);
+		const currentBranch = execSync("git rev-parse --abbrev-ref HEAD", {
+			cwd: worktreePath,
+		})
+			.toString()
+			.trim();
+		expect(currentBranch).toBe("feature/no-marker");
+	}, 10_000);
+
+	test("continues when git is killed mid-hook (timeout shape) but worktree is created", async () => {
```
[the remark this answers is not in the document]
Comment (Kitenite) [reply apps/desktop/src/lib/trpc/routers/workspaces/utils/git.test.ts:468]: Fixed in cf2bd257b: the test now asserts the tolerance warning actually fired (console.warn spy on the "non-fatal" message), so a trivially clean add — e.g. the signal never landing — fails the test instead of passing silently. If git's cleanup-on-signal behavior ever changes to remove the worktree, the test fails loudly, which is the correct signal: the tolerance depends on git leaving completed checkouts in place.

File: apps/desktop/src/lib/trpc/routers/workspaces/utils/git.ts [hunk 5]
```diff
@@ -1818,46 +1858,57 @@ export async function createWorktreeFromPr({
...
-					localBranchName,
-					"--no-track",
-					"FETCH_HEAD",
-				],
-				{ timeout: 30_000 },
-			);
-		}
+		// Checkouts run post-checkout hooks too — tolerate a hook failure when
+		// the PR branch verifiably ended up checked out.
+		await checkoutBranchWithHookTolerance({
```
[the remark this answers is not in the document]
Comment (Kitenite) [reply apps/desktop/src/lib/trpc/routers/workspaces/utils/git.ts]: Valid — fixed in bb8dc58df. The PR-checkout tolerance now requires HEAD to equal the PR's `headRefOid` (added to `gh pr view --json`) in addition to the branch match. A genuine gh failure (auth, network, failed fetch) leaves HEAD at the stale commit, so it rethrows; the error is only demoted when the checkout verifiably reached the PR's head. When the OID is unavailable, the error always surfaces — old behavior preserved.

File: apps/desktop/src/lib/trpc/routers/workspaces/utils/git.ts [hunk 6]
```diff
@@ -142,27 +142,63 @@ async function isWorktreeRegistered({
...
+	/**
+	 * Branch the worktree must have checked out for a failed add to count as
+	 * success. When omitted (detached adds), only a worktree registered by
+	 * THIS call qualifies — a path registered beforehand belongs to some
+	 * other checkout and its "already exists" failure must surface.
+	 */
+	expectedBranch?: string;
 	timeout?: number;
 }): Promise<void> {
+	const wasRegisteredBefore =
```
[the remark this answers is not in the document]
Comment (Kitenite) [reply apps/desktop/src/lib/trpc/routers/workspaces/utils/git.ts:175]: Agreed and intentional — documented in bb8dc58df on the `expectedBranch` JSDoc. A detached leftover is indistinguishable from a foreign checkout, and running `gh pr checkout --force` into the latter would clobber it, so detached strandings keep surfacing "already exists" (same as pre-PR behavior — this was never recoverable before either). Branch-add strandings, which are the common case (drew's), do recover. Making stranded PR retries self-heal safely needs a provenance marker (e.g. recording the worktree we registered) — noted as a follow-up.
